### ISSUE TITLE
Bank holiday dates subtract by 1 every time "save changes" is clicked.

### ISSUE THREAD

**Issue by coreymgilmore** [timeoff-management/application#107]
Details:
Timezone: EST.
Browser: Chrome.

So I was updating the Bank Holidays that our company takes off for.  I noticed the days were not "sticking" and kept decreasing.  Removed all holidays, started with just one.  Every time I click "Save changes" the date changed.

Went into the code and started digging around...  Ended up in `settings.js`.  Inserted a `console.log(company.bank_holiday);` at line 43 to see what was coming out of the DB.  Sure enough, dates are correct.  Threw a raw output in `general_settings.hbs` by putting 
`
{{#each company.bank_holidays}
{{name}} || {{date}}
{{/each}}
`
at the end of the file.  Output in browser is wrong (one day short).  Therefore every time "save changes" is clicked, the app (as it should) saves the new (one day previous) date.

Looks like this is caused by the parsing of the datetime coming from the db into the local timezone when the page is built.  Easy test for me....I just open up Dev Tools and create a new Date object and set the hours to 5 (since EST is -5 from GMT).  See below:

`new Date("2016-12-22T00:00:00.000Z");` = Wed Dec 21 2016 19:00:00 GMT-0500 (Eastern Standard Time)
versus:
`new Date("2016-12-22T05:00:00.000");` = Thu Dec 22 2016 00:00:00 GMT-0500 (Eastern Standard Time)

So clearly it is based on how the datetimes are stored in the db (in GMT timezone instead of local timezone) or displayed (converted from GMT to local).  Either way, one needs to be fixed.

NOTE: This also affect requesting time off.  For example: if I choose to take of from March 1st until March 4th, this is displayed as a request for Feb. 28 until Mar. 3.

Steps to reproduce:
1) Go to general settings.
2) Have at least one bank holiday created.
3) Look at the date.
4) Hit "Save changes."
5) Watch the page reload and the date decrease by one.

**Comment by coreymgilmore** [comment]
I am working on a fix but haven't come up with one yet.  I want something all encompassing.

Possibly a "timezone" setting in "General Settings" that localizes all dates to where the company is located?

Possibly some way to save the datetimes in the DB as a localized timezone instead of as GMT?

**Comment by vpp** [comment]
hi @coreymgilmore, thank you for such detailed description.

This is a problem indeed and there is a ticket in pipe line to address this (#58). I tried to crack on it but it is quite fundamental one so I decided to do one step at the time. Some preparation work was done and already in master (done as part of customisable date format).

The application assumes that it works with UTC time. So quick work around for you would be to set the time to be UTC. Although it is not the best solution as the app will be only usable for you between 5 in the morning and 7 in evening.

This issue is very serious and is the main reason why the app is not popular on the left side of Greenwich :)

I am considering this as high priority at this moment (even so some of work related to it is done as part of other tickets.)

**Comment by coreymgilmore** [comment]
Possible solutions:
 - store datetimes in DB per the local timezone instead of using UTC.
 - Use JS to localize all times after page is loaded (put a class on anything that is a date and on page load read the date, localize it, write it back to the element).

**Comment by coreymgilmore** [comment]
Quick and dirty fix....change the column type from "DATE" to "STRING" in db_design.txt and "DataTypes.DATE" to "DataTypes.STRING" in leave.js and bank_holiday.js.

It's not pretty(in the sense that dates aren't being stored in the db as dates) and it may cause other issues....but it works as of my testing.

Continuing to work on this and test...

**Comment by jamescobalt** [comment]
*I'm not a technical person. 
I think I misunderstood #169. Since the ical feed didn't work right due to the timezone difference, I assigned the local timezone to the server environment - but now I'm thinking maybe #169 was suggesting to remove the local timezone and set it to UTC +0? Regardless, this resulted in a number of errors that left the app inoperable - another person in here with one of the same errors said they simply started from scratch, so that's what I did. The system works again, but now #150 occurs. And FWIW, the mocha test is throwing a bunch of errors related to date behavior and does not complete. I didn't run the test before changing the server's timezone so I'm not sure how many of these are related to that change. I might try @coreymgilmore's hack if it hasn't proven problematic...
